File: ruby/lib/jewelmusic/exceptions.rb
# frozen_string_literal: true

module JewelMusic
  ##
  # Base exception for all JewelMusic API errors
  #
  class ApiError < StandardError
    attr_reader :status_code, :details

    ##
    # Initialize a new API error
    #
    # @param message [String] Error message
    # @param status_code [Integer] HTTP status code
    # @param details [Hash] Additional error details
    #
    def initialize(message = '', status_code = 0, details = {})
      super(message)
      @status_code = status_code
      @details = details
    end

    ##
    # Check if this is a client error (4xx)
    #
    # @return [Boolean]
    #
    def client_error?
      status_code >= 400 && status_code < 500
    end

    ##
    # Check if this is a server error (5xx)
    #
    # @return [Boolean]
    #
    def server_error?
      status_code >= 500
    end

    ##
    # Get error details as JSON string
    #
    # @return [String]
    #
    def details_json
      JSON.pretty_generate(details)
    end
  end

  ##
  # Exception raised when authentication fails
  #
  class AuthenticationError < ApiError
    def initialize(message = 'Authentication failed', status_code = 401, details = {})
      super(message, status_code, details)
    end
  end

  ##
  # Exception raised when a resource is not found
  #
  class NotFoundError < ApiError
    def initialize(message = 'Resource not found', status_code = 404, details = {})
      super(message, status_code, details)
    end
  end

  ##
  # Exception raised when request validation fails
  #
  class ValidationError < ApiError
    attr_reader :validation_errors

    def initialize(message = 'Validation failed', status_code = 422, validation_errors = {})
      super(message, status_code, validation_errors)
      @validation_errors = validation_errors
    end

    ##
    # Check if a specific field has validation errors
    #
    # @param field [String, Symbol] Field name
    # @return [Boolean]
    #
    def field_error?(field)
      validation_errors.key?(field.to_s) || validation_errors.key?(field.to_sym)
    end

    ##
    # Get validation errors for a specific field
    #
    # @param field [String, Symbol] Field name
    # @return [Array]
    #
    def field_errors(field)
      validation_errors[field.to_s] || validation_errors[field.to_sym] || []
    end
  end

  ##
  # Exception raised when rate limit is exceeded
  #
  class RateLimitError < ApiError
    attr_reader :retry_after

    def initialize(message = 'Rate limit exceeded', status_code = 429, retry_after = nil)
      super(message, status_code, {})
      @retry_after = retry_after
    end

    ##
    # Check if retry after time is specified
    #
    # @return [Boolean]
    #
    def retry_after?
      !retry_after.nil?
    end
  end

  ##
  # Exception raised for network and connection errors
  #
  class ConnectionError < ApiError
    def initialize(message = 'Connection error', original_error = nil)
      super(message, 0, { original_error: original_error })
    end
  end

  ##
  # Exception raised for request timeout errors
  #
  class TimeoutError < ApiError
    def initialize(message = 'Request timeout', timeout = nil)
      super(message, 0, { timeout: timeout })
    end
  end
end

File: python/jewelmusic_sdk/exceptions.py
"""
JewelMusic SDK Exception Classes

This module defines the exception hierarchy for the JewelMusic Python SDK.
All exceptions inherit from JewelMusicError for easy exception handling.
"""

from typing import Optional, Dict, Any


class JewelMusicError(Exception):
    """
    Base exception class for all JewelMusic SDK errors.
    
    All other SDK exceptions inherit from this class, allowing for
    easy exception handling with a single except clause.
    
    Attributes:
        message: Error message
        status_code: HTTP status code (if applicable)
        request_id: Request ID for debugging
        details: Additional error details
    """
    
    def __init__(
        self,
        message: str,
        status_code: Optional[int] = None,
        request_id: Optional[str] = None,
        details: Optional[Dict[str, Any]] = None
    ):
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.request_id = request_id
        self.details = details or {}
    
    def __str__(self) -> str:
        parts = [self.message]
        if self.request_id:
            parts.append(f"Request ID: {self.request_id}")
        if self.status_code:
            parts.append(f"Status: {self.status_code}")
        return " | ".join(parts)


class AuthenticationError(JewelMusicError):
    """
    Raised when API authentication fails (HTTP 401).
    
    This typically indicates:
    - Invalid API key
    - Expired API key
    - Missing authorization header
    """
    pass


class AuthorizationError(JewelMusicError):
    """
    Raised when the user lacks permission for the requested resource (HTTP 403).
    
    This typically indicates:
    - Insufficient API key permissions
    - Account limitations
    - Resource access restrictions
    """
    pass


class NotFoundError(JewelMusicError):
    """
    Raised when the requested resource is not found (HTTP 404).
    
    This typically indicates:
    - Invalid resource ID
    - Deleted resource
    - Incorrect endpoint URL
    """
    pass


class ValidationError(JewelMusicError):
    """
    Raised when request validation fails (HTTP 400).
    
    This typically indicates:
    - Invalid input parameters
    - Missing required fields
    - Data format errors
    - File validation failures
    
    Attributes:
        validation_errors: Detailed validation error information
    """
    
    def __init__(
        self,
        message: str,
        validation_errors: Optional[Dict[str, Any]] = None,
        **kwargs
    ):
        super().__init__(message, **kwargs)
        self.validation_errors = validation_errors or {}


class RateLimitError(JewelMusicError):
    """
    Raised when API rate limits are exceeded (HTTP 429).
    
    This indicates too many requests have been made in a given time period.
    
    Attributes:
        retry_after: Seconds to wait before retrying
        limit: Rate limit threshold
        remaining: Remaining requests in current window
        reset_time: When the rate limit window resets
    """
    
    def __init__(
        self,
        message: str,
        retry_after: Optional[int] = None,
        limit: Optional[int] = None,
        remaining: Optional[int] = None,
        reset_time: Optional[int] = None,
        **kwargs
    ):
        super().__init__(message, **kwargs)
        self.retry_after = retry_after
        self.limit = limit
        self.remaining = remaining
        self.reset_time = reset_time


class ServerError(JewelMusicError):
    """
    Raised when the server encounters an internal error (HTTP 5xx).
    
    This typically indicates:
    - Server-side processing errors
    - Database connectivity issues
    - Third-party service failures
    """
    pass


class NetworkError(JewelMusicError):
    """
    Raised when network connectivity issues occur.
    
    This typically indicates:
    - Connection timeouts
    - DNS resolution failures
    - Network connectivity issues
    - SSL/TLS errors
    """
    pass


class UnknownError(JewelMusicError):
    """
    Raised for unexpected errors that don't fit other categories.
    
    This is a catch-all for errors that don't have a specific exception type.
    """
    pass


class FileValidationError(ValidationError):
    """
    Raised when file validation fails.
    
    This typically indicates:
    - Unsupported file format
    - File size limits exceeded
    - Corrupted file data
    - Missing file metadata
    """
    pass


class ProcessingError(JewelMusicError):
    """
    Raised when audio or content processing fails.
    
    This typically indicates:
    - Audio processing failures
    - AI model errors
    - Content generation failures
    - Analysis processing errors
    """
    pass


class UploadError(JewelMusicError):
    """
    Raised when file upload operations fail.
    
    This typically indicates:
    - Upload timeout
    - Connection interruption
    - Server storage issues
    - File corruption during upload
    """
    pass


class ConfigurationError(JewelMusicError):
    """
    Raised when SDK configuration is invalid.
    
    This typically indicates:
    - Missing required configuration
    - Invalid configuration values
    - Environment setup issues
    """
    pass


def create_error_from_response(response_data: Dict[str, Any], status_code: int) -> JewelMusicError:
    """
    Create an appropriate exception from API response data.
    
    Args:
        response_data: API error response data
        status_code: HTTP status code
        
    Returns:
        Appropriate exception instance
    """
    message = response_data.get('error', {}).get('message', 'Unknown error')
    request_id = response_data.get('meta', {}).get('requestId')
    details = response_data.get('error', {}).get('details', {})
    
    error_kwargs = {
        'message': message,
        'status_code': status_code,
        'request_id': request_id,
        'details': details
    }
    
    if status_code == 400:
        return ValidationError(
            validation_errors=details.get('validation'),
            **error_kwargs
        )
    elif status_code == 401:
        return AuthenticationError(**error_kwargs)
    elif status_code == 403:
        return AuthorizationError(**error_kwargs)
    elif status_code == 404:
        return NotFoundError(**error_kwargs)
    elif status_code == 429:
        rate_limit_data = response_data.get('meta', {}).get('rateLimit', {})
        return RateLimitError(
            retry_after=rate_limit_data.get('reset'),
            limit=rate_limit_data.get('limit'),
            remaining=rate_limit_data.get('remaining'),
            reset_time=rate_limit_data.get('reset'),
            **error_kwargs
        )
    elif 500 <= status_code < 600:
        return ServerError(**error_kwargs)
    else:
        return UnknownError(**error_kwargs)